### PR TITLE
Bug 1184966 - perfherder should let harness do summarization

### PR DESCRIPTION
**Pull request by jmaher**
support for:
* old format
* current format (suite: val, subtests: {page: val, page: val}
* new format {suite: val, subtests: {page: {min: val, max: val, mean: val, median: val, std: val}...}}

and tests to validate the raw data.

[Review on Reviewable](https://reviewable.io/reviews/mozilla/treeherder/836)

### CODE REVIEW

File: treeherder/etl/perf_data_adapters.py [hunk 1]
```diff
@@ -90,16 +90,19 @@ def _round(num):
         return round(num, 2)
 
     @staticmethod
-    def _calculate_summary_data(job_id, result_set_id, push_timestamp, results):
+    def _calculate_summary_data(job_id, result_set_id, push_timestamp,
+                                results, summary):
```
Comment (wlach) [inline treeherder/etl/perf_data_adapters.py]: I don't passing a summary to this method is quite right -- I think we should just use the summary inline if it's available, and keep this method focused on calculating the summary where needed.

File: treeherder/etl/perf_data_adapters.py [hunk 2]
```diff
@@ -110,7 +113,7 @@ def _calculate_summary_data(job_id, result_set_id, push_timestamp, results):
 
     @staticmethod
     def _calculate_test_data(job_id, result_set_id, push_timestamp,
-                             replicates):
+                             replicates, summary):
```
Comment (wlach) [inline treeherder/etl/perf_data_adapters.py]: Likewise, I'd rather keep this method focused on calculating stuff where it's still needed. If necessary create a new method to transform the summary data provided by talos into the format we want and then conditionally call it instead. I think that'll save a lot of if/else logic, as well as being less confusing genrally.

File: tests/etl/test_perf_data_adapters.py [hunk 3]
```diff
@@ -50,4 +51,22 @@ def test_adapt_and_load():
         datum['blob'] = json.dumps({'talos_data': [datum['blob']]})
         tda.adapt_and_load(reference_data, job_data, datum)
 
+        # we upload a summary with a suite and subtest values, +1 for suite
+        if 'summary' in datum['blob']:
```
Comment (wlach) [inline tests/etl/test_perf_data_adapters.py]: I think this condition will always be true with your patch? Ideally we want to test both codepaths: one where the talos blob has a summary, one where it doesn't/

Comment (jmaher) [inline tests/etl/test_perf_data_adapters.py]: in this case I have modified some of the data in the sample data so we can
test the original format and the new format.

File: treeherder/etl/perf_data_adapters.py [hunk 4]
```diff
@@ -90,16 +90,21 @@ def _round(num):
         return round(num, 2)
 
     @staticmethod
-    def _calculate_summary_data(job_id, result_set_id, push_timestamp, results):
+    def _extract_summary_data(suite_data, summary):
+        suite_data["geomean"] = summary["suite"]
```
Comment (wlach) [inline treeherder/etl/perf_data_adapters.py]: We should file a followup bug to change the terminology here, since the suite summary value is not always a geometric mean. Maybe we should just say "value".

File: treeherder/etl/perf_data_adapters.py [hunk 5]
```diff
@@ -109,6 +114,26 @@ def _calculate_summary_data(job_id, result_set_id, push_timestamp, results):
         }
 
     @staticmethod
+    def _extract_test_data(series_data, summary):
+        if isinstance(summary, dict) and "min" in summary:
```
Comment (wlach) [inline treeherder/etl/perf_data_adapters.py]: I don't think it's necessary to check that summary is a dictionary all the time

File: tests/etl/test_perf_data_adapters.py (hunk 3 above)
Comment (wlach) [inline tests/etl/test_perf_data_adapters.py]: I see, I guess this is o.k. I would at least add an:

    else:
         # FIXME: the talos data blob we're currently using contains datums with summaries and those without
         # we should probably test non-summarized data as well
         pass

after the if clause to make it more clear what's happening

File: treeherder/etl/perf_data_adapters.py [hunk 6]
```diff
@@ -90,16 +90,21 @@ def _round(num):
         return round(num, 2)
 
     @staticmethod
-    def _calculate_summary_data(job_id, result_set_id, push_timestamp, results):
+    def _extract_summary_data(suite_data, summary):
+        suite_data["geomean"] = summary["suite"]
+        return suite_data
+
+    @staticmethod
+    def _calculate_summary_data(job_id, result_set_id, push_timestamp,
```
Comment (wlach) [inline treeherder/etl/perf_data_adapters.py]: It would be nice to avoid unnecessary changes here. I think you can do that just by pushing `results` up to the previous line and reverting the geomean changes below.

**Comment by wlach** [comment]
This has been merged manually.
